### PR TITLE
MAINT: Remove np.array_equal calls in tests

### PR DESCRIPTION
**Pull request by gfyoung**
Preferable to use `tm.assert_numpy_array_equal` instead.

### CODE REVIEW

**Comment by gfyoung** [comment]
A couple of questions going forward:

1) There are several instances where they used `np.array_equal` to check *non-array inequality*.  Would it be useful to create a method called `tm.assert_numpy_array_not_equal` for this case?

2) `tm.assert_numpy_array_equal` enforces that we must be comparing two `numpy` arrays, which makes it slightly cumbersome at times to compare two arrays (e.g. they are lists).  Perhaps we could take a page out of `numpy` book and create a function called `tm.assert_array_equal` ?

**Review by jreback — CHANGES_REQUESTED** [review]
need to add ``np.array_equal`` to the black-list in ci/lint.sh (near the end), we check for using black-listed functions and fail the build.

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 1]
```diff
@@ -22,7 +22,10 @@
 
 def eq_gen_range(kwargs, expected):
     rng = generate_range(**kwargs)
-    assert (np.array_equal(list(rng), expected))
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: I would remove this function and just directly compare these (only 3 uses)

File: pandas/tests/io/parser/test_textreader.py [hunk 2]
```diff
@@ -161,9 +161,9 @@ def test_skip_bad_lines(self):
                             error_bad_lines=False,
                             warn_bad_lines=False)
         result = reader.read()
-        expected = {0: ['a', 'd', 'g', 'l'],
-                    1: ['b', 'e', 'h', 'm'],
-                    2: ['c', 'f', 'i', 'n']}
+        expected = {0: np.array(['a', 'd', 'g', 'l'], dtype=object),
```
Comment (jreback) [inline pandas/tests/io/parser/test_textreader.py]: I think assert can do this; don't convert to a numpy array just to compare

File: pandas/tests/io/parser/test_textreader.py [hunk 3]
```diff
@@ -225,7 +225,7 @@ def test_escapechar(self):
         reader = TextReader(StringIO(data), delimiter=',', header=None,
                             escapechar='\\')
         result = reader.read()
-        expected = {0: ['"hello world"'] * 3}
+        expected = {0: np.array(['"hello world"'] * 3, dtype=object)}
```
Comment (jreback) [inline pandas/tests/io/parser/test_textreader.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by gfyoung** [comment]
> need to add `np.array_equal` to the black-list in ci/lint.sh (near the end)

There are many instances of this method being called in the source code...I think it shouldn't be allowed in tests BUT be careful with source code though.

**Comment by jreback** [comment]
> There are many instances of this method being called in the source code...I think it shouldn't be allowed in tests BUT be careful with source code though.

that line in lint should be

``grep -r -E --include '*.py' --exclude testing.py '(numpy|np)\.testing' pandas\tests``
(and have to add in the check for array_equal in the regex)

**Comment by jreback** [comment]
e.g.
```
(pandas) bash-3.2$   grep -r -E --include '*.py' --exclude testing.py '(numpy|np)\.testing|(numpy|np)\.array_equal' pandas/tests
pandas/tests/dtypes/test_cast.py:        assert (np.array_equal(result, arr))
pandas/tests/dtypes/test_cast.py:        assert (np.array_equal(result, expected))
```

**Comment by jreback** [comment]
@gfyoung can you do [this](https://github.com/pandas-dev/pandas/pull/18047#issuecomment-340880213) so we can ensure these don't appear going forward

**Comment by gfyoung** [comment]
@jreback : There are several instances that I did not remove yet because I want to do a more thorough refactoring in light of my comment <a href="https://github.com/pandas-dev/pandas/pull/18047#issuecomment-340659137">above</a>

I'll do that in a follow-up PR if that works.

File: pandas/tests/indexes/datetimes/test_date_range.py (hunk 1 above)
Comment (gfyoung) [reply pandas/tests/indexes/datetimes/test_date_range.py]: Done.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
that sounds good @gfyoung thanks!

File: pandas/tests/io/parser/test_textreader.py (hunk 2 above)
Comment (gfyoung) [reply pandas/tests/io/parser/test_textreader.py]: The output of `Textreader.read` is an array, so would need to convert.

File: pandas/tests/io/parser/test_textreader.py (hunk 3 above)
Comment (gfyoung) [reply pandas/tests/io/parser/test_textreader.py]: Same as <a href="https://github.com/pandas-dev/pandas/pull/18047#discussion_r148582611">above</a>

**Pull request merged**